pkg/controller/remotemachineset/gcpactuator.go

```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+	for _, l := range leases.Items {
+		logger.Debugf("checking lease: %s", l.Name)
+		for _, ownerRef := range l.OwnerReferences {
+			logger.Debugf("checking owner ref: %v", ownerRef)
+			if ownerRef.APIVersion == "hive.openshift.io/v1" &&
+				ownerRef.Kind == "MachinePool" &&
+				ownerRef.Name == pool.Name {
+
+				logger.Debugf("machine pool already has lease: %s", l.Name)
+
+				// Ensure the lease name is in the format we expect, we know everything up to
+				// the last character.
+				expectedLeasePrefix := fmt.Sprintf("%s-", cd.Spec.ClusterMetadata.InfraID)
+				if l.Name[0:len(l.Name)-1] != expectedLeasePrefix {
+					return "", false, fmt.Errorf("lease %s did not match expected lease name format (%s[CHAR])", l.Name, expectedLeasePrefix)
+				}
+
+				return string(l.Name[len(l.Name)-1]), false, nil
+			}
+		}
+	}
+	logger.Debugf("machine pool does not have a lease yet")
+
+	var leaseRune rune
+	// If the pool.Spec.Name == "worker", we want to preseve this MachinePool's "w" character that
+	// the installer would have selected so we do not cycle all worker nodes.
+	// Despite the separation of pool.Name and pool.Spec.Name, we do know that only one pool will
+	// have pool.Spec.Name worker as we validate that the pool must be named
+	// [clusterdeploymentname]-[pool.spec.name]
+	if pool.Spec.Name == "worker" {
+		leaseRune = 'w'
+		logger.Debug("selecting lease char 'w' for original worker pool")
+	} else {
+		// Pool does not have a lease yet, lookup all currently available lease chars
+		availLeaseChars, err := a.findAvailableLeaseChars(cd, leases)
+		if err != nil {
+			return "", false, err
+		}
+		logger.Debugf("availLeaseChars: %v", availLeaseChars)
+
+		// Choose a random entry in the available chars to limit collisions while processing
+		// multiple machine pools at the same time. In this case the subsequent attempts to create
+		// the lease will fail due to name collisions and re-reconcile.
+		logger.Debug("selecting random lease char from available")
+		leaseRune = availLeaseChars[rand.Intn(len(availLeaseChars))]
+	}
+	logger.Debugf("selected lease char: %s", string(leaseRune))
+
+	leaseName := fmt.Sprintf("%s-%s", cd.Spec.ClusterMetadata.InfraID, string(leaseRune))
+	// Attempt to claim the lease:
+	l := &hivev1.MachinePoolNameLease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      leaseName,
+			Namespace: pool.Namespace,
+			Labels: map[string]string{
+				constants.MachinePoolNameLabel:       pool.Name,
+				constants.ClusterDeploymentNameLabel: cd.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "hive.openshift.io/v1",
+					Kind:       "MachinePool",
+					Name:       pool.Name,
+					UID:        pool.UID,
+					Controller: pointer.BoolPtr(true),
+				},
+			},
+		},
+	}
+	logger.Debug("adding expectation for lease creation for this pool")
+	a.expectations.ExpectCreations(types.NamespacedName{Namespace: pool.Namespace, Name: pool.Name}.String(), 1)
```

Comment:
It smells to me a bit that the responsibility for expectations is split between the controller and the actuator. I can live with it for now, but I reserve the right to revisit it at some point.

Reply:
Yeah the thought process here was making all of this, leases and expectations, available to another actuator if they wanted to make use of it. I'm open to changing it in future though as I'm not sure this is the best either.